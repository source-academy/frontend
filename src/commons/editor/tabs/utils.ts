import path from 'path';

/**
 * Returns the shortest file paths that is uniquely identifiable among
 * all open editor tabs. This is similar to how most code editors available
 * handle the displaying of file names.
 *
 * For example, if there are 3 open editor tabs where the file names are
 * exactly the same, we would need to display more of the file path to be
 * able to distinguish which editor tab corresponds to which file. Given
 * the following absolute file paths:
 * - /a.js
 * - /dir1/a.js
 * - /dir1/dir2/a.js
 * The shortest unique file paths will be:
 * - /a.js
 * - /dir1/a.js
 * - /dir2/a.js
 *
 * @param originalFilePaths The file paths to be shortened.
 */
export const getShortestUniqueFilePaths = (originalFilePaths: string[]): string[] => {
  // Store the unique shortened file paths as a mapping from the original file paths
  // to the shortened file paths. This is necessary because the output of this function
  // must preserve the original ordering of file paths.
  const originalToUniqueShortenedFilePaths: Record<string, string> = {};
  // Split each original file path into path segments and store the mapping from file
  // path to path segments for O(1) lookup. Since we only deal with the BrowserFS file
  // system, the path separator will always be '/'.
  const filePathSegments: Record<string, string[]> = originalFilePaths.reduce(
    (segments, filePath) => ({
      ...segments,
      // It is necessary to remove empty segments to deal with the very first '/' in
      // file paths.
      [filePath]: filePath.split(path.sep).filter(segment => segment !== '')
    }),
    {}
  );

  for (
    let numOfPathSegments = 1;
    // Keep looping while some original file paths have yet to be shortened.
    Object.keys(originalToUniqueShortenedFilePaths).length < originalFilePaths.length;
    numOfPathSegments++
  ) {
    // Based on the number of path segments for the iteration, we construct the
    // shortened file path. We then store the mapping from the shortened file path
    // to any original file path which transforms into it.
    const shortenedToOriginalFilePaths: Record<string, string[]> = Object.entries(
      filePathSegments
    ).reduce((filePaths, [originalFilePath, filePathSegments]) => {
      // Note that if there are fewer path segments than the number being sliced,
      // all of the path segments will be returned without error.
      const shortenedFilePath = '/' + filePathSegments.slice(-numOfPathSegments).join('/');
      return {
        ...filePaths,
        [shortenedFilePath]: (filePaths[shortenedFilePath] ?? []).concat(originalFilePath)
      };
    }, {});
    // Each shortened file path that only has a single corresponding original file
    // path is added to the unique shortened file paths record and their entry in
    // the file path segments record is removed to prevent further processing.
    Object.entries(shortenedToOriginalFilePaths).forEach(
      ([shortenedFilePath, originalFilePaths]) => {
        if (originalFilePaths.length > 1) {
          return;
        }

        const originalFilePath = originalFilePaths[0];
        originalToUniqueShortenedFilePaths[originalFilePath] = shortenedFilePath;
        // Remove the file path's segments from the next iteration.
        delete filePathSegments[originalFilePath];
      }
    );
  }

  // Finally, we retrieve the unique shortened file paths while preserving the ordering
  // of file paths.
  return originalFilePaths.map(filePath => originalToUniqueShortenedFilePaths[filePath]);
};
