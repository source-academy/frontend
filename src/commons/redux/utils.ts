import {
  ActionCreatorWithOptionalPayload,
  ActionCreatorWithoutPayload,
  ActionCreatorWithPreparedPayload,
  createAction
} from '@reduxjs/toolkit';
import * as Sentry from '@sentry/browser';
import { SagaIterator } from 'redux-saga';
import { StrictEffect, takeEvery } from 'redux-saga/effects';

/**
 * Creates actions, given a base name and base actions
 * @param baseName The base name of the actions
 * @param baseActions The base actions. Use a falsy value to create an action without a payload.
 * @returns An object containing the actions
 */
export function createActions<BaseName extends string, BaseActions extends Record<string, any>>(
  baseName: BaseName,
  baseActions: BaseActions
) {
  return Object.entries(baseActions).reduce(
    (res, [name, func]) => ({
      ...res,
      [name]: func
        ? createAction(`${baseName}/${name}`, (...args: any) => ({ payload: func(...args) }))
        : createAction(`${baseName}/${name}`)
    }),
    {} as {
      [K in keyof BaseActions]: K extends string
        ? BaseActions[K] extends (...args: any) => any
          ? ActionCreatorWithPreparedPayload<
              Parameters<BaseActions[K]>,
              ReturnType<BaseActions[K]>,
              `${BaseName}/${K}`
            >
          : ActionCreatorWithoutPayload<`${BaseName}/${K}`>
        : never;
    }
  );
}

export function combineSagaHandlers<
  TActions extends Record<
    string,
    ActionCreatorWithPreparedPayload<any, any> | ActionCreatorWithoutPayload<any>
  >
>(
  actions: TActions,
  handlers: {
    // TODO: Maybe this can be stricter? And remove the optional type after
    // migration is fully done
    [K in keyof TActions]?: (action: ReturnType<TActions[K]>) => SagaIterator;
  },
  others?: (takeEvery: typeof saferTakeEvery) => SagaIterator
): () => SagaIterator {
  const sagaHandlers = Object.entries(handlers).map(([actionName, saga]) =>
    saferTakeEvery(actions[actionName], saga)
  );
  return function* (): SagaIterator {
    yield* sagaHandlers;
    if (others) {
      const obj = others(saferTakeEvery);
      while (true) {
        const { done, value } = obj.next();
        if (done) break;
        yield value;
      }
    }
  };
}

export function saferTakeEvery<
  Action extends
    | ActionCreatorWithOptionalPayload<any>
    | ActionCreatorWithPreparedPayload<any[], any>
>(actionPattern: Action, fn: (action: ReturnType<Action>) => Generator<StrictEffect<any>>) {
  function* wrapper(action: ReturnType<Action>) {
    try {
      yield* fn(action);
    } catch (error) {
      handleUncaughtError(error);
    }
  }

  return takeEvery(actionPattern.type, wrapper);
}

function handleUncaughtError(error: any) {
  if (process.env.NODE_ENV === 'development') {
    // react-error-overlay is a "special" package that's automatically included
    // in development mode by CRA

    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
    // @ts-ignore
    import('react-error-overlay').then(reo => reo.reportRuntimeError(error));
  }
  Sentry.captureException(error);
  console.error(error);
}
