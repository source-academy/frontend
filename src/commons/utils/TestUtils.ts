import React from 'react';
import renderer from 'react-test-renderer';
import { createRenderer } from 'react-test-renderer/shallow';

export const shallowRender = (element: React.ReactElement) => {
  const renderer = createRenderer();
  renderer.render(element);
  return renderer.getRenderOutput();
};

export const renderTreeJson = (element: React.ReactElement) => {
  return renderer.create(element).toJSON();
};

export const renderTree = (element: React.ReactElement) => {
  return renderer.create(element);
};

/**
 * Recursively traverses a nested object and returns all matches.
 *
 * Used in traversing nested objects such as `shallowRender`'s output.
 */
export function deepFilter<T>(
  nestedObject: T,
  matchFn: (e: T) => boolean,
  getChildren: (e: T) => T[]
) {
  const matches: any[] = [];

  function helper(obj: T) {
    if (matchFn(obj)) {
      matches.push(obj);
    }

    const children = getChildren(obj);
    if (children && Array.isArray(children)) {
      children.forEach(e => {
        if (e) {
          helper(e);
        }
      });
    }
  }

  helper(nestedObject);
  return matches;
}
