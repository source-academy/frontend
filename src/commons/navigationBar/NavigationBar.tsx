import {
  Alignment,
  Button,
  Classes,
  Drawer,
  FocusStyleManager,
  Icon,
  Navbar,
  NavbarDivider,
  NavbarGroup,
  NavbarHeading,
  Position
} from '@blueprintjs/core';
import { IconNames } from '@blueprintjs/icons';
import { BlueprintIcons_16Id } from '@blueprintjs/icons/lib/esm/generated-icons/16px/blueprint-icons-16';
import { Popover2 } from '@blueprintjs/popover2';
import classNames from 'classnames';
import { Location } from 'history';
import { useCallback, useMemo, useState } from 'react';
import { NavLink, Route, Routes, useLocation } from 'react-router-dom';

import Dropdown from '../dropdown/Dropdown';
import NotificationBadge from '../notificationBadge/NotificationBadge';
import { filterNotificationsByType } from '../notificationBadge/NotificationBadgeHelper';
import Constants from '../utils/Constants';
import { useResponsive, useSession } from '../utils/Hooks';
import AcademyNavigationBar, {
  assessmentTypesToNavlinkInfo,
  getAcademyNavbarRightInfo
} from './subcomponents/AcademyNavigationBar';
import NavigationBarLangSelectButton from './subcomponents/NavigationBarLangSelectButton';
import SicpNavigationBar from './subcomponents/SicpNavigationBar';

export type NavbarEntryInfo = {
  to: string;
  icon: BlueprintIcons_16Id;
  text: string;
  disabled?: boolean; // entry is not rendered when disabled
  hasNotifications?: boolean; // whether to render NotificationBadge
  hiddenInBreakpoints?: ('xs' | 'sm' | 'md' | 'lg')[]; // hide text in Blueprint breakpoints
};

type CreateNavlinkFunction = (navbarEntry: NavbarEntryInfo) => React.ReactElement;

const NavigationBar: React.FC = () => {
  const [mobileSideMenuOpen, setMobileSideMenuOpen] = useState(false);
  const { isMobileBreakpoint } = useResponsive();
  const location = useLocation();
  const {
    isLoggedIn,
    isEnrolledInACourse,
    role,
    courseId,
    courseShortName,
    enableAchievements,
    enableSourcecast,
    assessmentConfigurations
  } = useSession();
  const assessmentTypes = useMemo(
    () => assessmentConfigurations?.map(c => c.type),
    [assessmentConfigurations]
  );

  FocusStyleManager.onlyShowFocusOnTabs();

  const createMobileNavlink: CreateNavlinkFunction = useCallback(
    navbarEntry => (
      <NavLink
        to={navbarEntry.to}
        className={({ isActive }) =>
          classNames(
            'NavigationBar__link__mobile',
            Classes.BUTTON,
            Classes.MINIMAL,
            Classes.LARGE,
            {
              [Classes.ACTIVE]: isActive
            }
          )
        }
        onClick={() => setMobileSideMenuOpen(false)}
        key={navbarEntry.text}
      >
        <Icon icon={navbarEntry.icon} />
        <div>{navbarEntry.text}</div>
        {navbarEntry.hasNotifications && (
          <NotificationBadge
            notificationFilter={filterNotificationsByType(navbarEntry.text)}
            disableHover={true}
          />
        )}
      </NavLink>
    ),
    [setMobileSideMenuOpen]
  );

  const wrapWithMobileHamburger = (navlinks: (React.ReactElement | null)[]) => {
    // Don't render drawer when there are 0 navlinks in it
    const nonNullNavlinks = navlinks.filter(e => e !== null);
    const renderDrawer = nonNullNavlinks.length > 0;

    return (
      <NavbarGroup align={Alignment.LEFT}>
        {renderDrawer && (
          <Button
            onClick={() => setMobileSideMenuOpen(!mobileSideMenuOpen)}
            icon={IconNames.MENU}
            large={true}
            minimal={true}
          />
        )}
        <NavLink
          className={classNames('NavigationBar__link', Classes.BUTTON, Classes.MINIMAL)}
          to={
            Constants.playgroundOnly ? '/' : courseId == null ? '/welcome' : `/courses/${courseId}`
          }
        >
          <Icon icon={IconNames.SYMBOL_DIAMOND} />
          <NavbarHeading style={{ paddingBottom: '0px' }}>
            {courseShortName || Constants.sourceAcademyDeploymentName}
          </NavbarHeading>
        </NavLink>
        {renderDrawer && (
          <Drawer
            isOpen={mobileSideMenuOpen}
            position="left"
            onClose={() => setMobileSideMenuOpen(false)}
            title=""
            className={Classes.DARK}
            style={{ overflowY: 'auto' }}
          >
            {navlinks}
          </Drawer>
        )}
      </NavbarGroup>
    );
  };

  const fullAcademyNavbarLeftAssessmentsInfo: NavbarEntryInfo[] = useMemo(
    () =>
      assessmentTypesToNavlinkInfo({
        assessmentTypes,
        courseId,
        isEnrolledInACourse
      }),
    [assessmentTypes, courseId, isEnrolledInACourse]
  );

  const fullAcademyNavbarLeftCommonInfo: NavbarEntryInfo[] = useMemo(() => {
    return [
      {
        to: `/courses/${courseId}/sourcecast`,
        icon: IconNames.MUSIC,
        text: 'Sourcecast',
        disabled: !(isEnrolledInACourse && enableSourcecast)
      },
      {
        to: '/playground',
        icon: IconNames.CODE,
        text: 'Playground',
        disabled: !isEnrolledInACourse
      },
      {
        to: '/githubassessments',
        icon: IconNames.BRIEFCASE,
        text: 'Classroom',
        disabled: !Constants.enableGitHubAssessments
      },
      {
        to: '/sicpjs',
        icon: IconNames.BOOK,
        text: 'SICP JS',
        disabled: !isLoggedIn
      },
      {
        to: `/courses/${courseId}/achievements`,
        icon: IconNames.MOUNTAIN,
        text: 'Achievements',
        disabled: !(isEnrolledInACourse && enableAchievements)
      },
      {
        to: '/stories',
        icon: IconNames.GIT_REPO,
        text: 'Stories',
        // TODO: Enable when stories are implemented
        disabled: true && !isLoggedIn
      }
    ];
  }, [isLoggedIn, isEnrolledInACourse, courseId, enableSourcecast, enableAchievements]);

  const fullAcademyMobileNavbarLeftAdditionalInfo = useMemo(
    () => getAcademyNavbarRightInfo({ isEnrolledInACourse, courseId, role }),
    [isEnrolledInACourse, courseId, role]
  );

  const fullAcademyMobileNavbarLeftInfoWithAssessments: NavbarEntryInfo[] = useMemo(() => {
    return [
      ...fullAcademyNavbarLeftAssessmentsInfo,
      ...fullAcademyNavbarLeftCommonInfo,
      ...fullAcademyMobileNavbarLeftAdditionalInfo
    ];
  }, [
    fullAcademyNavbarLeftAssessmentsInfo,
    fullAcademyNavbarLeftCommonInfo,
    fullAcademyMobileNavbarLeftAdditionalInfo
  ]);

  const renderPlaygroundOnlyNavbarLeftDesktop = () => (
    <NavbarGroup align={Alignment.LEFT}>
      {renderNavlinksFromInfo(playgroundOnlyNavbarLeftInfo, createDesktopNavlink)}
    </NavbarGroup>
  );

  const renderPlaygroundOnlyNavbarLeftMobile = () =>
    wrapWithMobileHamburger(
      renderNavlinksFromInfo(playgroundOnlyNavbarLeftInfo, createMobileNavlink)
    );

  const renderFullAcademyNavbarLeftDesktop = () => {
    const desktopNavbarLeftPopoverContent = (
      <Navbar>
        <NavbarGroup>
          {renderNavlinksFromInfo(
            assessmentTypesToNavlinkInfo({
              assessmentTypes,
              courseId,
              isEnrolledInACourse
            }),
            createDesktopNavlink
          )}
        </NavbarGroup>
      </Navbar>
    );
    const topNavbarNavlinks = [
      '/playground',
      '/sicpjs',
      '/contributors',
      '/githubassessments',
      `/courses/${courseId}/sourcecast`,
      `/courses/${courseId}/achievements`
    ];
    const enableDesktopPopover =
      courseId != null && !!topNavbarNavlinks.find(x => location.pathname.startsWith(x));
    const highlightDesktopLogo = (location: Location) => {
      // Highlight main logo when none of the topmost-blue navbar links are active
      return !topNavbarNavlinks.find(x => location.pathname.startsWith(x));
    };

    return (
      <NavbarGroup align={Alignment.LEFT}>
        <Popover2
          position={Position.BOTTOM_RIGHT}
          interactionKind="hover"
          content={desktopNavbarLeftPopoverContent}
          popoverClassName={'desktop-navbar-popover'}
          disabled={!enableDesktopPopover}
        >
          <NavLink
            className={classNames('NavigationBar__link', Classes.BUTTON, Classes.MINIMAL, {
              [Classes.ACTIVE]: highlightDesktopLogo(location)
            })}
            to={courseId == null ? '/welcome' : `/courses/${courseId}`}
          >
            <Icon icon={IconNames.SYMBOL_DIAMOND} />
            <NavbarHeading style={{ paddingBottom: '0px' }}>
              {courseShortName || Constants.sourceAcademyDeploymentName}
            </NavbarHeading>
          </NavLink>
        </Popover2>
        {renderNavlinksFromInfo(fullAcademyNavbarLeftCommonInfo, createDesktopNavlink)}
      </NavbarGroup>
    );
  };

  const renderFullAcademyNavbarLeftMobile = () =>
    wrapWithMobileHamburger(
      renderNavlinksFromInfo(fullAcademyMobileNavbarLeftInfoWithAssessments, createMobileNavlink)
    );

  const commonNavbarRight = (
    <NavbarGroup align={Alignment.RIGHT}>
      {location.pathname.startsWith('/playground') && <NavigationBarLangSelectButton />}
      <NavLink
        className={({ isActive }) =>
          classNames('NavigationBar__link', Classes.BUTTON, Classes.MINIMAL, {
            [Classes.ACTIVE]: isActive
          })
        }
        to="/contributors"
      >
        <Icon icon={IconNames.HEART} />
      </NavLink>

      <div className="visible-xs">
        <NavbarDivider className="thin-divider" />
      </div>
      <div className="hidden-xs">
        <NavbarDivider className="default-divider" />
      </div>

      <Dropdown />
    </NavbarGroup>
  );

  return (
    <>
      <Navbar className={classNames('NavigationBar', 'primary-navbar', Classes.DARK)}>
        {Constants.playgroundOnly
          ? isMobileBreakpoint
            ? renderPlaygroundOnlyNavbarLeftMobile()
            : renderPlaygroundOnlyNavbarLeftDesktop()
          : isMobileBreakpoint
          ? renderFullAcademyNavbarLeftMobile()
          : renderFullAcademyNavbarLeftDesktop()}
        {commonNavbarRight}
      </Navbar>

      <Routes>
        <Route path="/playground" element={null} />
        <Route path="/githubassessments/*" element={null} />
        <Route path="/contributors" element={null} />
        <Route path="/courses/:courseId/sourcecast" element={null} />
        <Route path="/courses/:courseId/achievements" element={null} />
        <Route path="/sicpjs/:section?" element={<SicpNavigationBar />} />
        <Route
          path="*"
          element={
            !Constants.playgroundOnly && isEnrolledInACourse && !isMobileBreakpoint ? (
              <AcademyNavigationBar assessmentTypes={assessmentTypes} />
            ) : null
          }
        />
      </Routes>
    </>
  );
};

const playgroundOnlyNavbarLeftInfo: NavbarEntryInfo[] = [
  {
    to: '/playground',
    icon: IconNames.CODE,
    text: 'Playground'
  },
  {
    to: '/githubassessments',
    icon: IconNames.BRIEFCASE,
    text: 'Classroom',
    disabled: !Constants.enableGitHubAssessments
  },
  {
    to: '/sicpjs',
    icon: IconNames.BOOK,
    text: 'SICP JS'
  },
  {
    to: '/stories',
    icon: IconNames.GIT_REPO,
    text: 'Stories',
    // TODO: Enable when stories are implemented
    disabled: true
  }
];

export const renderNavlinksFromInfo = (
  navbarEntries: NavbarEntryInfo[],
  createNavlink: CreateNavlinkFunction
): (React.ReactElement | null)[] =>
  navbarEntries.map(entry => {
    if (entry.disabled) {
      return null;
    }

    return createNavlink(entry);
  });

export const createDesktopNavlink: CreateNavlinkFunction = navbarEntry => (
  <NavLink
    className={({ isActive }) =>
      classNames('NavigationBar__link', Classes.BUTTON, Classes.MINIMAL, {
        [Classes.ACTIVE]: isActive
      })
    }
    to={navbarEntry.to}
    key={navbarEntry.text}
    title={navbarEntry.text}
  >
    <Icon icon={navbarEntry.icon} />
    <div
      className={classNames(
        'navbar-button-text',
        navbarEntry.hiddenInBreakpoints?.map(bp => `hidden-${bp}`)
      )}
    >
      {navbarEntry.text}
    </div>
    {navbarEntry.hasNotifications && (
      <NotificationBadge
        notificationFilter={filterNotificationsByType(navbarEntry.text)}
        disableHover={true}
      />
    )}
  </NavLink>
);

export default NavigationBar;
