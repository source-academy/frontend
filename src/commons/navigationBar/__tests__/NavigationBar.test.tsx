import { useLocation } from 'react-router';
import { useTypedSelector } from 'src/commons/utils/Hooks';
import { shallowRender } from 'src/commons/utils/TestUtils';
import { Mock, vi } from 'vitest';

import { Role } from '../../application/ApplicationTypes';
import NavigationBar from '../NavigationBar';

vi.mock('react-router', async () => ({
  ...(await vi.importActual('react-router')),
  useLocation: vi.fn()
}));
vi.mock('react-redux', async () => ({
  ...(await vi.importActual('react-redux')),
  useSelector: vi.fn()
}));

const useSelectorMock = useTypedSelector as Mock;
const useLocationMock = useLocation as Mock;

describe('NavigationBar', () => {
  beforeEach(() => {
    useLocationMock.mockReturnValue({
      pathname: 'localhost:8000/courses/1/game'
    });
  });

  it('Renders "Not logged in" correctly', () => {
    useSelectorMock.mockReturnValueOnce({});
    const tree = shallowRender(<NavigationBar />);
    expect(tree).toMatchSnapshot();
  });

  it('Renders correctly for student with course', () => {
    useSelectorMock.mockReturnValueOnce({
      role: Role.Student,
      name: 'Bob',
      courseId: 1,
      courseShortName: 'CS1101S',
      enableAchievements: true,
      enableContestLeaderboard: false,
      enableOverallLeaderboard: false,
      enableSourcecast: true,
      assessmentConfigurations: [
        {
          type: 'Missions'
        },
        {
          type: 'Quests'
        },
        {
          type: 'Paths'
        },
        {
          type: 'Contests'
        },
        {
          type: 'Missions'
        }
      ]
    });

    const tree = shallowRender(<NavigationBar />);
    expect(tree).toMatchSnapshot();
  });

  test('Renders correctly for student without course', () => {
    useSelectorMock.mockReturnValueOnce({
      name: 'Bob'
    });
    const tree = shallowRender(<NavigationBar />);
    expect(tree).toMatchSnapshot();
  });
});
