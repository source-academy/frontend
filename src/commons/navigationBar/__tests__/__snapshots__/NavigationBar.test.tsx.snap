// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`NavigationBar > Renders "Not logged in" correctly 1`] = `
<React.Fragment>
  <Blueprint6.Navbar
    className="NavigationBar primary-navbar _primary-navbar_d292d6 bp6-dark"
  >
    <Blueprint6.NavbarGroup
      align="left"
    >
      <Blueprint6.Popover
        boundary="clippingParents"
        captureDismiss={false}
        content={
          <Blueprint6.Navbar>
            <Blueprint6.NavbarGroup />
          </Blueprint6.Navbar>
        }
        defaultIsOpen={false}
        disabled={true}
        fill={false}
        hasBackdrop={false}
        hoverCloseDelay={300}
        hoverOpenDelay={150}
        inheritDarkTheme={true}
        interactionKind="hover"
        matchTargetWidth={false}
        minimal={false}
        openOnTargetFocus={true}
        popoverClassName="desktop-navbar-popover"
        position="bottom-right"
        positioningStrategy="absolute"
        shouldReturnFocusOnClose={false}
        targetTagName="span"
        transitionDuration={300}
        usePortal={true}
      >
        <NavLink
          className="NavigationBar__link"
          to="/welcome"
        >
          <Blueprint6.NavbarHeading>
            <Blueprint6.Button
              active={true}
              className="app-title"
              icon="symbol-diamond"
              minimal={true}
            >
              Source Academy @ NUS
            </Blueprint6.Button>
          </Blueprint6.NavbarHeading>
        </NavLink>
      </Blueprint6.Popover>
      <DesktopNavLink
        disabled={true}
        icon="music"
        text="Sourcecast"
        to="/courses/undefined/sourcecast"
      />
      <DesktopNavLink
        disabled={true}
        icon="code"
        text="Playground"
        to="/playground"
      />
      <DesktopNavLink
        disabled={true}
        icon="book"
        text="SICP JS"
        to="/sicpjs"
      />
      <DesktopNavLink
        disabled={true}
        icon="mountain"
        text="Achievements"
        to="/courses/undefined/achievements"
      />
      <DesktopNavLink
        disabled={true}
        icon="git-repo"
        text="Stories"
        to="/courses/undefined/stories"
      />
      <DesktopNavLink
        disabled={true}
        icon="timeline-bar-chart"
        text="Leaderboard"
        to="/courses/undefined/leaderboard"
      />
    </Blueprint6.NavbarGroup>
    <Blueprint6.NavbarGroup
      align="right"
    >
      <NavigationBarLangSelectButton />
      <NavLink
        className={[Function]}
        to="/contributors"
      >
        <Blueprint6.Icon
          icon="heart"
        />
      </NavLink>
      <div
        className="visible-xs"
      >
        <Blueprint6.NavbarDivider
          className="thin-divider"
        />
      </div>
      <div
        className="hidden-xs"
      >
        <Blueprint6.NavbarDivider
          className="default-divider"
        />
      </div>
      <Dropdown />
    </Blueprint6.NavbarGroup>
  </Blueprint6.Navbar>
  <Routes>
    <Route
      element={null}
      path="/playground/*"
    />
    <Route
      element={null}
      path="/contributors"
    />
    <Route
      element={null}
      path="/courses/:courseId/sourcecast"
    />
    <Route
      element={null}
      path="/courses/:courseId/achievements"
    />
    <Route
      element={null}
      path="/courses/:courseId/leaderboard/*"
    />
    <Route
      element={<SicpNavigationBar />}
      path="/sicpjs/:section?"
    />
    <Route
      element={null}
      path="*"
    />
  </Routes>
</React.Fragment>
`;

exports[`NavigationBar > Renders correctly for student with course 1`] = `
<React.Fragment>
  <Blueprint6.Navbar
    className="NavigationBar primary-navbar _primary-navbar_d292d6 bp6-dark"
  >
    <Blueprint6.NavbarGroup
      align="left"
    >
      <Blueprint6.Popover
        boundary="clippingParents"
        captureDismiss={false}
        content={
          <Blueprint6.Navbar>
            <Blueprint6.NavbarGroup>
              <DesktopNavLink
                disabled={false}
                hasNotifications={true}
                hiddenInBreakpoints={
                  [
                    "xs",
                    "sm",
                  ]
                }
                icon="flame"
                text="Missions"
                to="/courses/1/missions"
              />
              <DesktopNavLink
                disabled={false}
                hasNotifications={true}
                hiddenInBreakpoints={
                  [
                    "xs",
                    "sm",
                  ]
                }
                icon="lightbulb"
                text="Quests"
                to="/courses/1/quests"
              />
              <DesktopNavLink
                disabled={false}
                hasNotifications={true}
                hiddenInBreakpoints={
                  [
                    "xs",
                    "sm",
                  ]
                }
                icon="predictive-analysis"
                text="Paths"
                to="/courses/1/paths"
              />
              <DesktopNavLink
                disabled={false}
                hasNotifications={true}
                hiddenInBreakpoints={
                  [
                    "xs",
                    "sm",
                  ]
                }
                icon="comparison"
                text="Contests"
                to="/courses/1/contests"
              />
              <DesktopNavLink
                disabled={false}
                hasNotifications={true}
                hiddenInBreakpoints={
                  [
                    "xs",
                    "sm",
                  ]
                }
                icon="manual"
                text="Missions"
                to="/courses/1/missions"
              />
            </Blueprint6.NavbarGroup>
          </Blueprint6.Navbar>
        }
        defaultIsOpen={false}
        disabled={true}
        fill={false}
        hasBackdrop={false}
        hoverCloseDelay={300}
        hoverOpenDelay={150}
        inheritDarkTheme={true}
        interactionKind="hover"
        matchTargetWidth={false}
        minimal={false}
        openOnTargetFocus={true}
        popoverClassName="desktop-navbar-popover"
        position="bottom-right"
        positioningStrategy="absolute"
        shouldReturnFocusOnClose={false}
        targetTagName="span"
        transitionDuration={300}
        usePortal={true}
      >
        <NavLink
          className="NavigationBar__link"
          to="/courses/1"
        >
          <Blueprint6.NavbarHeading>
            <Blueprint6.Button
              active={true}
              className="app-title"
              icon="symbol-diamond"
              minimal={true}
            >
              CS1101S
            </Blueprint6.Button>
          </Blueprint6.NavbarHeading>
        </NavLink>
      </Blueprint6.Popover>
      <DesktopNavLink
        disabled={false}
        icon="music"
        text="Sourcecast"
        to="/courses/1/sourcecast"
      />
      <DesktopNavLink
        disabled={false}
        icon="code"
        text="Playground"
        to="/playground"
      />
      <DesktopNavLink
        disabled={false}
        icon="book"
        text="SICP JS"
        to="/sicpjs"
      />
      <DesktopNavLink
        disabled={false}
        icon="mountain"
        text="Achievements"
        to="/courses/1/achievements"
      />
      <DesktopNavLink
        disabled={true}
        icon="git-repo"
        text="Stories"
        to="/courses/1/stories"
      />
      <DesktopNavLink
        disabled={true}
        icon="timeline-bar-chart"
        text="Leaderboard"
        to="/courses/1/leaderboard"
      />
    </Blueprint6.NavbarGroup>
    <Blueprint6.NavbarGroup
      align="right"
    >
      <NavigationBarLangSelectButton />
      <NavLink
        className={[Function]}
        to="/contributors"
      >
        <Blueprint6.Icon
          icon="heart"
        />
      </NavLink>
      <div
        className="visible-xs"
      >
        <Blueprint6.NavbarDivider
          className="thin-divider"
        />
      </div>
      <div
        className="hidden-xs"
      >
        <Blueprint6.NavbarDivider
          className="default-divider"
        />
      </div>
      <Dropdown />
    </Blueprint6.NavbarGroup>
  </Blueprint6.Navbar>
  <Routes>
    <Route
      element={null}
      path="/playground/*"
    />
    <Route
      element={null}
      path="/contributors"
    />
    <Route
      element={null}
      path="/courses/:courseId/sourcecast"
    />
    <Route
      element={null}
      path="/courses/:courseId/achievements"
    />
    <Route
      element={null}
      path="/courses/:courseId/leaderboard/*"
    />
    <Route
      element={<SicpNavigationBar />}
      path="/sicpjs/:section?"
    />
    <Route
      element={
        <AcademyNavigationBar
          assessmentTypes={
            [
              "Missions",
              "Quests",
              "Paths",
              "Contests",
              "Missions",
            ]
          }
        />
      }
      path="*"
    />
  </Routes>
</React.Fragment>
`;

exports[`NavigationBar > Renders correctly for student without course 1`] = `
<React.Fragment>
  <Blueprint6.Navbar
    className="NavigationBar primary-navbar _primary-navbar_d292d6 bp6-dark"
  >
    <Blueprint6.NavbarGroup
      align="left"
    >
      <Blueprint6.Popover
        boundary="clippingParents"
        captureDismiss={false}
        content={
          <Blueprint6.Navbar>
            <Blueprint6.NavbarGroup />
          </Blueprint6.Navbar>
        }
        defaultIsOpen={false}
        disabled={true}
        fill={false}
        hasBackdrop={false}
        hoverCloseDelay={300}
        hoverOpenDelay={150}
        inheritDarkTheme={true}
        interactionKind="hover"
        matchTargetWidth={false}
        minimal={false}
        openOnTargetFocus={true}
        popoverClassName="desktop-navbar-popover"
        position="bottom-right"
        positioningStrategy="absolute"
        shouldReturnFocusOnClose={false}
        targetTagName="span"
        transitionDuration={300}
        usePortal={true}
      >
        <NavLink
          className="NavigationBar__link"
          to="/welcome"
        >
          <Blueprint6.NavbarHeading>
            <Blueprint6.Button
              active={true}
              className="app-title"
              icon="symbol-diamond"
              minimal={true}
            >
              Source Academy @ NUS
            </Blueprint6.Button>
          </Blueprint6.NavbarHeading>
        </NavLink>
      </Blueprint6.Popover>
      <DesktopNavLink
        disabled={true}
        icon="music"
        text="Sourcecast"
        to="/courses/undefined/sourcecast"
      />
      <DesktopNavLink
        disabled={true}
        icon="code"
        text="Playground"
        to="/playground"
      />
      <DesktopNavLink
        disabled={false}
        icon="book"
        text="SICP JS"
        to="/sicpjs"
      />
      <DesktopNavLink
        disabled={true}
        icon="mountain"
        text="Achievements"
        to="/courses/undefined/achievements"
      />
      <DesktopNavLink
        disabled={true}
        icon="git-repo"
        text="Stories"
        to="/courses/undefined/stories"
      />
      <DesktopNavLink
        disabled={true}
        icon="timeline-bar-chart"
        text="Leaderboard"
        to="/courses/undefined/leaderboard"
      />
    </Blueprint6.NavbarGroup>
    <Blueprint6.NavbarGroup
      align="right"
    >
      <NavigationBarLangSelectButton />
      <NavLink
        className={[Function]}
        to="/contributors"
      >
        <Blueprint6.Icon
          icon="heart"
        />
      </NavLink>
      <div
        className="visible-xs"
      >
        <Blueprint6.NavbarDivider
          className="thin-divider"
        />
      </div>
      <div
        className="hidden-xs"
      >
        <Blueprint6.NavbarDivider
          className="default-divider"
        />
      </div>
      <Dropdown />
    </Blueprint6.NavbarGroup>
  </Blueprint6.Navbar>
  <Routes>
    <Route
      element={null}
      path="/playground/*"
    />
    <Route
      element={null}
      path="/contributors"
    />
    <Route
      element={null}
      path="/courses/:courseId/sourcecast"
    />
    <Route
      element={null}
      path="/courses/:courseId/achievements"
    />
    <Route
      element={null}
      path="/courses/:courseId/leaderboard/*"
    />
    <Route
      element={<SicpNavigationBar />}
      path="/sicpjs/:section?"
    />
    <Route
      element={null}
      path="*"
    />
  </Routes>
</React.Fragment>
`;
